### PR TITLE
- Added Puppet 4 types for all parameters.

### PR DESCRIPTION
**Pull request by rehanone**
- Added Puppet 4 types for all parameters.
- Converted the module to use Puppet Development Kit.
- Fixed style errors.
- Fixed naming errors.
- Updated minimum puppet and dependencies versions and dropped unsupported versions.

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -1,38 +1,27 @@
...
+  Boolean $service_manage           = $autofs::params::service_manage,
+  Boolean $service_hasrestart       = $autofs::params::service_hasrestart,
+  Boolean $service_hasstatus        = $autofs::params::service_hasstatus,
+  Enum[running, stopped]
+  $service_ensure                   = $autofs::params::service_ensure,
+  Boolean $service_enable           = $autofs::params::service_enable,
+  Optional[String] $service_restart = $autofs::params::service_restart,
+  Hash $mapfiles                    = lookup('autofs::mapfiles', Hash, 'hash', {}),
+  Hash $mounts                      = lookup('autofs::mounts', Hash, 'hash', {}),
+  Hash $includes                    = lookup('autofs::includes', Hash, 'hash', {}),
```
Comment (jkroepke) [inline manifests/init.pp]: Puppet provides an automatic hiera lookup for all class parameters. This isn't necessary

File: manifests/include.pp [hunk 2]
```diff
@@ -1,16 +1,17 @@
-define autofs::include(
-  $mapfile = $name,
-  $order   = undef
+#
+define autofs::include (
+  String $mapfile         = $title,
```
Comment (jkroepke) [inline manifests/include.pp]: Why do you prefer title instead name?

File: manifests/init.pp [hunk 3]
```diff
@@ -1,38 +1,27 @@
...
-  $includes           = {},
+class autofs (
+  String $config_file_owner         = $autofs::params::config_file_owner,
+  String $config_file_group         = $autofs::params::config_file_group,
+  String $master_config             = $autofs::params::master_config,
+  Stdlib::Absolutepath
+  $map_config_dir                   = $autofs::params::map_config_dir,
+  Boolean $package_manage           = $autofs::params::package_manage,
+  Array[String] $package_name       = $autofs::params::package_name,
+  Enum[present, absent, purged, installed]
```
Comment (jkroepke) [inline manifests/init.pp]: String should be good enough, because ensure allows to define a specific version.

File: manifests/init.pp [hunk 4]
```diff
@@ -1,38 +1,27 @@
...
+  $package_ensure                   = $autofs::params::package_ensure,
+  String $service_name              = $autofs::params::service_name,
+  Boolean $service_manage           = $autofs::params::service_manage,
+  Boolean $service_hasrestart       = $autofs::params::service_hasrestart,
+  Boolean $service_hasstatus        = $autofs::params::service_hasstatus,
+  Enum[running, stopped]
+  $service_ensure                   = $autofs::params::service_ensure,
```
Comment (jkroepke) [inline manifests/init.pp]: True and False are also allowed. https://puppet.com/docs/puppet/5.4/types/service.html#service-attribute-ensure

File: manifests/mapfile.pp [hunk 5]
```diff
@@ -1,42 +1,26 @@
...
-  $ensure   = present,
-  $mapfile  = $name,
-  $options  = undef,
-  $order    = undef,
-  $maptype  = 'file' ,
-  $mounts   = {}
+define autofs::mapfile (
+  Optional[Stdlib::Absolutepath] $directory,
+  Enum[present, absent, purged] $ensure = present,
+  String $mapfile                       = $title,
```
Comment (jkroepke) [inline manifests/mapfile.pp]: Why $title instead $name?

File: manifests/mapfile.pp [hunk 6]
```diff
@@ -1,42 +1,26 @@
...
+  Optional[String] $options             = undef,
+  Optional[String] $order               = undef,
+  Enum[ file, program, yp, nisplus, hesiod, ldap, ldaps]
+  $maptype                              = file,
```
Comment (jkroepke) [inline manifests/mapfile.pp]: No string quotas anymore required?

File: manifests/include.pp (hunk 2 above)
Comment (rehanone) [reply manifests/include.pp]: I was using this as a reference https://puppet.com/docs/puppet/5.5/lang_defined_types.html#title-and-name

But I think name is also good and I can revert it.

Comment (jkroepke) [reply manifests/include.pp]: Ah, nice to know. 

Since mapfile should be unique, $title is better then name.

File: manifests/mapfile.pp (hunk 6 above)
Comment (rehanone) [reply manifests/mapfile.pp]: https://puppet.com/docs/puppet/5.5/lang_data_string.html#bare-words

**Comment by rehanone** [comment]
@jkroepke Can you please check now.

**Pull request merged**